net/HttpRequest.hpp

```diff
@@ -1609,11 +1614,35 @@ class Session final : public ProtocolHandlerInterface
 
             if (!socket->send(_request))
             {
+                _result = net::asyncConnectResult::SocketError;
                 LOG_ERR("Error while writing to socket");
             }
         }
     }
 
+    void callOnConnectFail()
+    {
+        if (!_onConnectFail)
+            return;
+
+        auto self = shared_from_this();
+        try
+        {
+            [[maybe_unused]] const auto references = self.use_count();
+            assert(references > 1 && "Expected more than 1 reference to http::Session.");
+
+            _onConnectFail(std::static_pointer_cast<Session>(self));
```

Comment:
... here it would be more clear why you need a cast. I would expect this is Session already.

Reply:
self is of type `std::shared_ptr<ProtocolHandlerInterface> `, `shared_from_this` will only return super-class instance, here `ProtocolHandlerInterface`, I anc add an explicit `std::shared_ptr<Session> Session::shared_from_this` implementation.

Reply:
Ah, fair enough. Other code does this conversion in a dedicated function, see e.g. 'client_from_this'. Fine to do that in a followup PR.